docs/pallas/tpu/matmul.md

```diff
@@ -0,0 +1,643 @@
+
+In Pallas, we specify that via `BlockSpec`s and a `grid`. Note that we already have a nested for loop in the block matrix multiplication algorithm. We can specify that in Pallas via a `grid`. The slices in the block matrix multiplication can also be specified via `BlockSpec`s.
+
++++ {"id": "FvYoyqlyIqo6"}
+
+## Your first matrix multiplication kernel
+
++++ {"id": "umKZAlSvIt7x"}
+
+Putting it all together, here's an implementation of a block matrix multiplication kernel that pipelines the memory transfers with the compute. We create a 3-d grid, corresponding to the 3-nested loop in the NumPy code. Note that while MXUs are only capable of multiplying small blocks, Pallas will automatically take bigger blocks and automatically tile them over the MXUs.
+
+The last dimension of the grid corresponds to the contraction dimension of the matrix multiply and is a reduction dimension, so we need to be sure to initialize the accumulator.
+
+```{code-cell} ipython3
+:id: 75FBANKFbmQ5
+
+def matmul_kernel(x_ref, y_ref, z_ref):
+  @pl.when(pl.program_id(2) == 0)
+  def _():
+    z_ref[...] = jnp.zeros_like(z_ref)
+
+  z_ref[...] += x_ref[...] @ y_ref[...]
+
+def matmul(
+    x: jax.Array,
+    y: jax.Array,
+    *,
+    bm: int = 128,
+    bk: int = 128,
+    bn: int = 128,
+):
+  m, k = x.shape
+  _, n = y.shape
+  return pl.pallas_call(
+      matmul_kernel,
+      out_shape=jax.ShapeDtypeStruct((m, n), x.dtype),
+      in_specs=[pl.BlockSpec(lambda i, j, k: (i, k), (bm, bk)),
+                pl.BlockSpec(lambda i, j, k: (k, j), (bk, bn))],
+      out_specs=pl.BlockSpec(lambda i, j, k: (i, j), (bm, bn)),
+      grid=(m // bm, n // bn, k // bk)
+  )(x, y)
+```
+
+```{code-cell} ipython3
+:id: 0e8qTsimccGV
+
+m, k, n = 4096, 4096, 4096
+k1, k2 = random.split(random.key(0), 2)
+x = random.normal(k1, (m, k), dtype=jnp.float32)
+y = random.normal(k2, (k, n), dtype=jnp.float32)
+np.testing.assert_array_equal(x @ y, matmul(x, y))
+```
+
++++ {"id": "DycJX_-PJnnB"}
+
+## Matrix multiplication performance
+
+Let's think about how to analyze matrix multiplication performance.
+
+The number of FLOPs in a `(m, k) x (k, n)` matrix multiplication are (approximately) `2 * m * k * n`. (Technically it is `n * m * (2k - 1)` but for large enough `k` our approximation is sufficient.)
+
+The minimum amount of memory usage for a matrix multiply (assuming float32) is the total size of the inputs (copying into VMEM) plus the size of the output (copying into HBM). Thus the minimum bandwidth usage is `(m * k + k * n + m * n) * 4 bytes/float32`.
+
+One observation is that the number of matmul FLOPs is cubic in its inputs whereas the minimum bandwidth usage is quadratic in its inputs. Intuitively, this means that FLOPs grow faster than bandwidth usage, meaning that the bigger our matmul is, the more compute we have relative to copying.
```

Comment:
assuming we only read the inputs once, which cannot always be the case in practice